app/grandchallenge/notifications/utils.py

```diff
@@ -0,0 +1,141 @@
+import collections
+
+from django.contrib.contenttypes.fields import GenericForeignKey
+from django.contrib.contenttypes.models import ContentType
+from django.db.models.fields.related_descriptors import (
+    ForwardManyToOneDescriptor,
+)
+
+
+# Code in this file was adapted from https://djangosnippets.org/snippets/2492/
+# and https://andrew.hawker.io/writings/2020/11/18/django-activity-stream-prefetch-generic-foreign-key/
+
+
+def _queryset_foreign_keys(queryset, generic_fk):
+    fks = {}
+    for name, field in queryset.model.__dict__.items():
+        if not generic_fk and not isinstance(
+            field, ForwardManyToOneDescriptor
+        ):
+            continue
+        elif generic_fk and not isinstance(field, GenericForeignKey):
+            continue
+        fks[name] = field
+    return fks
+
+
+def _content_type_to_content_mapping_for_nested_gfks(queryset, fks):
+    data = collections.defaultdict(list)
+    gfks = {}
+    for weak_model in queryset:
+        for _, fk_field in fks.items():
+            for name, fk in fk_field.field.related_model.__dict__.items():
+                if not isinstance(fk, GenericForeignKey):
+                    continue
+                gfks[name] = fk
+                (
+                    related_content_type,
+                    related_object_id,
+                ) = _get_related_content_type_and_related_object_pk(
+                    weak_model.action, fk_field.field.related_model, fk
+                )
+                if related_content_type is None:
+                    continue
+                if related_object_id is None:
+                    continue
+
+                data[related_content_type].append(related_object_id)
+    return data, gfks
+
+
+def _content_type_to_content_mapping_for_gfks(queryset, gfks):
+    data = collections.defaultdict(list)
+    for (
+        _model,
+        _field_name,
+        content_type,
+        object_pk,
+    ) in _queryset_gfk_content_generator(queryset, gfks):
+        data[content_type].append(object_pk)
+    return data
+
+
+def _get_related_content_type_and_related_object_pk(model, fk_field, fk):
+    related_content_type_id = getattr(
+        model, fk_field._meta.get_field(fk.ct_field).get_attname(),
+    )
+    try:
+        related_object_id = int(getattr(model, fk.fk_field))
+        related_content_type = ContentType.objects.get_for_id(
+            related_content_type_id
+        )
+    except TypeError:
+        related_object_id = None
+        related_content_type = None
+
+    return related_content_type, related_object_id
+
+
+def _queryset_gfk_content_generator(queryset, gfks):
+    for model in queryset:
+        for field_name, field in gfks.items():
+            (
+                content_type,
+                object_pk,
+            ) = _get_related_content_type_and_related_object_pk(
+                model, field.model, field
+            )
+            yield (model, field_name, content_type, object_pk)

```

Comment:
Maybe use a named tuple here and unpack by name, not item. https://docs.quantifiedcode.com/python-anti-patterns/readability/not_using_named_tuples_when_returning_more_than_one_value.html